### ISSUE TITLE
RFC: TC39 observable compatibility

### ISSUE THREAD

**Issue by znerol** [Kotlin/kotlinx.coroutines#2207]
Reactive JavaScript frameworks are standardizing around [tc39/proposal-observable](https://github.com/tc39/proposal-observable). It would be nice to have a `Flow<T>` extension returning a suitable `SubscriberFunction`.

PoC thrown together rather quickly without considering any backpressure / coroutine scope issues.

```kotlin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch

/**
 * Turns a flow into a TC39 [SubscriberFunction] suitable for consumption by compliant reactive frameworks such as rxjs.
 *
 * https://github.com/tc39/proposal-observable
 *
 * Example (Kotlin):
 *     @JsExport
 *     val testSubscriber = flowOf(1,2,3,4).asSubscriber()
 *
 * Example (JavaScript):
 *     const obs = new Observer(testSubscriber);
 *     obs.subscribe(x => console.log(x));
 */
fun <T> Flow<T>.asSubscriber(scope: CoroutineScope = GlobalScope) : SubscriberFunction<T> = { observer : SubscriptionObserver<T> ->
    val job = scope.launch {
        try {
            collect { value ->
                if (!observer.closed) {
                    observer.next(value)
                }
            }
        }
        catch(e: Exception) {
            observer.error(e)
        }
        finally {
            observer.complete()
        }
    }

    object : Subscription {
        override fun unsubscribe() {
            job.cancel()
        }
        override val closed = job.isCompleted
    }
}

external interface SubscriptionObserver<T> {
    // Sends the next value in the sequence
    fun next(value: T)

    // Sends the sequence error
    fun error(errorValue: dynamic)

    // Sends the completion notification
    fun complete()

    // A boolean value indicating whether the subscription is closed
    val closed : Boolean
}

external interface Subscription {
    // Cancels the subscription
    fun unsubscribe(): Unit;

    // A boolean value indicating whether the subscription is closed
    val closed: Boolean;
}

typealias SubscriberFunction<T> = (observer: SubscriptionObserver<T>) -> Subscription
```

**Comment by fvasco** [comment]
Why not a third party integration?

**Comment by elizarov** [comment]
If this becomes an actual standard it would make sense to support it out-of-the-box directly in `kotlinx-coroutines-core`, just like JS Promise is supported.

**Comment by fvasco** [comment]
...in the meanwhile...

**Comment by znerol** [comment]
[znerol/kotlinx-coroutines-observable-js](https://github.com/znerol/kotlinx-coroutines-observable-js) very much work in progress. Tests are looking good when everything is compiled with `IR` on node and in the browser. When compiled with `LEGACY` tests in node pass while those in the browser fail -- for both `JavaScript/Webpack` and `TypeScript` builds.

**Comment by znerol** [comment]
Released 1.0.0 to bintray and jCenter.
